# Install Libraries

In [1]:
%pip install scikit-learn
%pip install pandas
%pip install matplotlib
%pip install seaborn
%pip install plotly
%pip install --upgrade nbformat

Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


Requirement already up-to-date: nbformat in .\env\lib\site-packages (5.9.2)
Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'd:\an-online-shop-business\env\Scripts\python.exe -m pip install --upgrade pip' command.


# Importing Libraries

In [2]:
# import sklearn
# from sklearn.linear_model import LinearRegression

# Importing for Data Manipulation
# import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Importing For Data Visualization
# import matplotlib.pyplot as plt # for data visualization
# import seaborn as sns
# from wordcloud import WordCloud
# import plotly.express as px
# import plotly.graph_objs as go
# from plotly.offline import init_notebook_mode, iplot

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

# import os
# for dirname, _, filenames in os.walk('/kaggle/input'):
#     for filename in filenames:
#         print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Reading The Dataset

In [3]:
# Reading the dataset with Pandas
df = pd.read_csv("./Sales Transaction v.4a.csv")

# Sampling the data based on heads (5 top ranks)
df.head()

TransactionNo       Date ProductNo                          ProductName  \
0        581482  12/9/2019     22485        Set Of 2 Wooden Market Crates   
1        581475  12/9/2019     22596  Christmas Star Wish List Chalkboard   
2        581475  12/9/2019     23235             Storage Tin Vintage Leaf   
3        581475  12/9/2019     23272    Tree T-Light Holder Willie Winkie   
4        581475  12/9/2019     23239    Set Of 4 Knick Knack Tins Poppies   

   Price  Quantity  CustomerNo         Country  
0  21.47        12     17490.0  United Kingdom  
1  10.65        36     13069.0  United Kingdom  
2  11.53        12     13069.0  United Kingdom  
3  10.65        12     13069.0  United Kingdom  
4  11.94         6     13069.0  United Kingdom

In [4]:
# Sampling the data based on tails (5 least ranks)
df.tail()

TransactionNo       Date ProductNo                   ProductName  \
536345       C536548  12/1/2018     22168  Organiser Wood Antique White   
536346       C536548  12/1/2018     21218        Red Spotty Biscuit Tin   
536347       C536548  12/1/2018     20957  Porcelain Hanging Bell Small   
536348       C536548  12/1/2018     22580  Advent Calendar Gingham Sack   
536349       C536548  12/1/2018     22767    Triple Photo Frame Cornice   

        Price  Quantity  CustomerNo  Country  
536345  18.96        -2     12472.0  Germany  
536346  14.09        -3     12472.0  Germany  
536347  11.74        -1     12472.0  Germany  
536348  16.35        -4     12472.0  Germany  
536349  20.45        -2     12472.0  Germany

# Understanding The Data

#### Based on the source of the data, we can describe the dataset to understanding the meaning of the data. Each data can be describe as follow :
 1. TransactionNo (categorical): a six-digit unique number that defines each transaction. The letter “C” in the code indicates a cancellation.

 2. Date (numeric): the date when each transaction was generated.

 3. ProductNo (categorical): a five or six-digit unique character used to identify a specific product.

 4. Product (categorical): product/item name.

 5. Price (numeric): the price of each product per unit in pound sterling (£).

 6. Quantity (numeric): the quantity of each product per transaction. Negative values related to cancelled transactions.

 7. CustomerNo (categorical): a five-digit unique number that defines each customer.

 8. Country (categorical): name of the country where the customer resides.


In [5]:
# Print the shape of the dataset
df.shape

(536350, 8)

In [6]:
# Print to describing about the dataset
df.describe()

Price       Quantity     CustomerNo
count  536350.000000  536350.000000  536295.000000
mean       12.662182       9.919347   15227.893178
std         8.490450     216.662300    1716.582932
min         5.130000  -80995.000000   12004.000000
25%        10.990000       1.000000   13807.000000
50%        11.940000       3.000000   15152.000000
75%        14.090000      10.000000   16729.000000
max       660.620000   80995.000000   18287.000000

In [7]:
# Print to look for info about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [8]:
# Print to searching about missing values
print ("Missing Values")
print ("-"*30)
print(df.isnull().sum())

Missing Values
------------------------------
TransactionNo     0
Date              0
ProductNo         0
ProductName       0
Price             0
Quantity          0
CustomerNo       55
Country           0
dtype: int64


In [9]:
# Searching for data duplicates
print ("Data Duplicatioin")
print ("-"*30)
print(df.duplicated(keep=False).sum())

Data Duplicatioin
------------------------------
9994


In [10]:
# Check the unique value on Dataset
print("Unique Values")
print("-"*30)
print(df.nunique())

Unique Values
------------------------------
TransactionNo    23204
Date               305
ProductNo         3768
ProductName       3768
Price              530
Quantity           510
CustomerNo        4738
Country             38
dtype: int64


In [11]:
print(f"There's {df['Country'].nunique()} Country in this Dataset")
print("-"*30)
print('Including :', df['Country'].unique())

There's 38 Country in this Dataset
------------------------------
Including : ['United Kingdom' 'Norway' 'Belgium' 'Germany' 'France' 'Austria'
 'Netherlands' 'EIRE' 'USA' 'Channel Islands' 'Iceland' 'Portugal' 'Spain'
 'Finland' 'Italy' 'Greece' 'Japan' 'Sweden' 'Denmark' 'Cyprus' 'Malta'
 'Switzerland' 'Australia' 'Czech Republic' 'Poland' 'Hong Kong'
 'Singapore' 'RSA' 'Israel' 'Unspecified' 'United Arab Emirates' 'Canada'
 'European Community' 'Bahrain' 'Brazil' 'Saudi Arabia' 'Lebanon'
 'Lithuania']


From here we learn that :

1. The dataset contain (536350 rows of transaction, 8 collumns of attributes).
2. The 'Date' collumn stored as an object (text), which is not a proper date and time format
3. The 'CustomerNo' collumn stored as an float numbers, which can cause ambiguousity
4. Total of country that this database have is 38 Country. Which mean this online shop business have transaction across the country in the world.
5. Some data have missing values, including the CustomerNo, which is unexpected. Every transaction mustly have a CustomerNo in there.
6. In Quantity collumn, there have negative values. It can cause ambigousity to.
7. In Transaction collumn, we must remember that if theres have letter "C" in the first value. That indicate cancellation.

# Data Cleaning

In [12]:
# Display the rows in the 'CostumerNo' collumn:
rows_in_costumerno_missing = df[df['CustomerNo'].isnull()]
print ("Rows CostumerNo With Missing Values")
rows_in_costumerno_missing.head()

Rows CostumerNo With Missing Values


TransactionNo        Date ProductNo  \
6511         C581406   12/8/2019    46000M   
6512         C581406   12/8/2019    46000S   
90098        C575153   11/8/2019     22947   
102671       C574288   11/3/2019     22178   
117263       C573180  10/28/2019     23048   

                                ProductName  Price  Quantity  CustomerNo  \
6511           Polyester Filler Pad 45x45cm   6.19      -240         NaN   
6512           Polyester Filler Pad 40x40cm   6.19      -300         NaN   
90098            Wooden Advent Calendar Red  44.25        -1         NaN   
102671      Victorian Glass Hanging T-Light  25.37        -1         NaN   
117263  Set Of 10 Lanterns Fairy Light Star  14.50        -1         NaN   

               Country  
6511    United Kingdom  
6512    United Kingdom  
90098   United Kingdom  
102671  United Kingdom  
117263  United Kingdom

In [13]:
#cleaning the data with missing values
df = df.dropna()
#checking the data
print ("Missing Values")
print ("-"*30)
print(df.isnull().sum())

Missing Values
------------------------------
TransactionNo    0
Date             0
ProductNo        0
ProductName      0
Price            0
Quantity         0
CustomerNo       0
Country          0
dtype: int64


In [14]:
#check if there's duplicates in dataset
print ('Data Duplicates')
print ("-"*30)
print (df.duplicated(keep=False).sum())

Data Duplicates
------------------------------
9994


In [15]:
#cleaning the duplicates
df = df.drop_duplicates()

In [16]:
#checking again if there's duplicates in dataset
print ('Data Duplicates')
print ("-"*30)
print (df.duplicated(keep=False).sum())

Data Duplicates
------------------------------
0


In [17]:
#check description of the databased
df.describe()

Price       Quantity     CustomerNo
count  531095.000000  531095.000000  531095.000000
mean       12.669635       9.993146   15222.612241
std         8.526181     217.710261    1716.633588
min         5.130000  -80995.000000   12004.000000
25%        10.990000       1.000000   13798.000000
50%        11.940000       3.000000   15146.000000
75%        14.090000      10.000000   16727.000000
max       660.620000   80995.000000   18287.000000

> if we look at the database, there have negative value in 'Quantity'. We must clean that
and make sure that TransactionNo must'nt have 'C' indicator in there (Cancellation).

Disini akan dilakukan pembagian data yaitu data yang terdapat pembatalan pembelian dan data yang tanpa pembatalan pembelian

In [18]:
# Searching the TransactionNo with 'C' indicators
df_cancel = df[df['TransactionNo'].str.startswith('C') & (df['Quantity'] < 0)]

# Print the cancelation rows
print ("Cancelation Orders")
df_cancel

Cancelation Orders


TransactionNo       Date ProductNo                       ProductName  \
1616         C581484  12/9/2019     23843         Paper Craft Little Birdie   
1617         C581490  12/9/2019     22178   Victorian Glass Hanging T-Light   
1618         C581490  12/9/2019     23144   Zinc T-Light Holder Stars Small   
1619         C581568  12/9/2019     21258        Victorian Sewing Box Large   
1620         C581569  12/9/2019     84978  Hanging Heart Jar T-Light Holder   
...              ...        ...       ...                               ...   
536345       C536548  12/1/2018     22168      Organiser Wood Antique White   
536346       C536548  12/1/2018     21218            Red Spotty Biscuit Tin   
536347       C536548  12/1/2018     20957      Porcelain Hanging Bell Small   
536348       C536548  12/1/2018     22580      Advent Calendar Gingham Sack   
536349       C536548  12/1/2018     22767        Triple Photo Frame Cornice   

        Price  Quantity  CustomerNo         Country  
1616     6.19    -80995     16446.0  United Kingdom  
1617     6.19       -12     14397.0  United Kingdom  
1618     6.04       -11     14397.0  United Kingdom  
1619     6.19        -5     15311.0  United Kingdom  
1620     6.19        -1     17315.0  United Kingdom  
...       ...       ...         ...             ...  
536345  18.96        -2     12472.0         Germany  
536346  14.09        -3     12472.0         Germany  
536347  11.74        -1     12472.0         Germany  
536348  16.35        -4     12472.0         Germany  
536349  20.45        -2     12472.0         Germany  

[8494 rows x 8 columns]

In [19]:
df_cancel.describe()

Price      Quantity    CustomerNo
count  8494.000000   8494.000000   8494.000000
mean     14.667710    -31.496586  14994.995526
std      25.241861   1197.896194   1706.171419
min       5.460000 -80995.000000  12346.000000
25%      11.340000     -6.000000  13520.750000
50%      12.860000     -2.000000  14901.500000
75%      15.320000     -1.000000  16393.000000
max     660.620000     -1.000000  18282.000000

In [20]:
# lalu kemudian nilai minus (-) diubah menjadi positif agar dapat dihitung nantinya
df_cancel['Quantity'] = df_cancel['Quantity'].abs()
df_cancel

C:\Users\Zarel\AppData\Local\Temp\ipykernel_23800\123655012.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cancel['Quantity'] = df_cancel['Quantity'].abs()


TransactionNo       Date ProductNo                       ProductName  \
1616         C581484  12/9/2019     23843         Paper Craft Little Birdie   
1617         C581490  12/9/2019     22178   Victorian Glass Hanging T-Light   
1618         C581490  12/9/2019     23144   Zinc T-Light Holder Stars Small   
1619         C581568  12/9/2019     21258        Victorian Sewing Box Large   
1620         C581569  12/9/2019     84978  Hanging Heart Jar T-Light Holder   
...              ...        ...       ...                               ...   
536345       C536548  12/1/2018     22168      Organiser Wood Antique White   
536346       C536548  12/1/2018     21218            Red Spotty Biscuit Tin   
536347       C536548  12/1/2018     20957      Porcelain Hanging Bell Small   
536348       C536548  12/1/2018     22580      Advent Calendar Gingham Sack   
536349       C536548  12/1/2018     22767        Triple Photo Frame Cornice   

        Price  Quantity  CustomerNo         Country  
1616     6.19     80995     16446.0  United Kingdom  
1617     6.19        12     14397.0  United Kingdom  
1618     6.04        11     14397.0  United Kingdom  
1619     6.19         5     15311.0  United Kingdom  
1620     6.19         1     17315.0  United Kingdom  
...       ...       ...         ...             ...  
536345  18.96         2     12472.0         Germany  
536346  14.09         3     12472.0         Germany  
536347  11.74         1     12472.0         Germany  
536348  16.35         4     12472.0         Germany  
536349  20.45         2     12472.0         Germany  

[8494 rows x 8 columns]

In [21]:
# Delete Cancelation Transaction based on Quantity below 0
df_surplus = df[df['Quantity'] > 0]

In [22]:
# Checking the databased again
df_surplus.describe()

Price       Quantity     CustomerNo
count  522601.000000  522601.000000  522601.000000
mean       12.637160      10.667492   15226.311767
std         7.965974     157.542420    1716.555479
min         5.130000       1.000000   12004.000000
25%        10.990000       1.000000   13804.000000
50%        11.940000       4.000000   15152.000000
75%        14.090000      12.000000   16729.000000
max       660.620000   80995.000000   18287.000000

In [23]:
df_surplus.info()

<class 'pandas.core.frame.DataFrame'>
Index: 522601 entries, 0 to 536324
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  522601 non-null  object 
 1   Date           522601 non-null  object 
 2   ProductNo      522601 non-null  object 
 3   ProductName    522601 non-null  object 
 4   Price          522601 non-null  float64
 5   Quantity       522601 non-null  int64  
 6   CustomerNo     522601 non-null  float64
 7   Country        522601 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 35.9+ MB


In [24]:
%store df_surplus
%store df_cancel

Stored 'df_surplus' (DataFrame)
Stored 'df_cancel' (DataFrame)
